Support multiple instance of localhost Multiplayer on 1 PC

anr2me opened the pull request:
I only briefly tested this on Win10 with Metal Slug XX, because the game i have on this laptop is only games with AdhocMatching so i can't test it on regular Adhoc, ~~and apparently AdhocMatching is broken in the latest version of PPSSPP (not sure since when it's broken), when i tried to join a host it gets frozen (deadlock/race-condition issue during AdhocMatching callbacks maybe?)~~

anr2me committed:
Support multiple instance of localhost Multiplayer on 1 PC

LunaMoo commented:
Hi, could you look at:
https://github.com/ANR2ME/ppsspp/compare/adhoc_fix...LunaMoo:AdhocStuff_from_THBranch

 That's a comparison between this branch of yours and mine which I separated from my hackish branch that had some of your old work merged and supposedly even not including hacks has a better compatibility for multiplier. Outside of hacks I only made some small changes like replacing all: 
````
real_time_now()*1000000.0
````
with 
````
CoreTiming::GetGlobalTimeUsScaled()
````

But maybe the rest of the differences is fixing the adhocMatching you mentioned that's broken in PPSSPP's current master. Feel free to salvage that code, it's yours anyway.

 You could also take my changes for README.md where I removed duplicate credits we already had in it's proper place and  UI/MiscScreens.cpp where I added you to the credits and aemu to the 3rd party tools used to keep it consistent with everything else and finally place your name there at the same time.

ANR2ME commented:
> 
> 
> Hi, could you look at:
> [ANR2ME/ppsspp@adhoc_fix...LunaMoo:AdhocStuff_from_THBranch](https://github.com/ANR2ME/ppsspp/compare/adhoc_fix...LunaMoo:AdhocStuff_from_THBranch)
> 
> That's a comparison between this branch of yours and mine which I separated from my hackish branch that had some of your old work merged and supposedly even not including hacks has a better compatibility for multiplier. Outside of hacks I only made some small changes like replacing all:
> 
> ```
> real_time_now()*1000000.0
> ```
> 
> with
> 
> ```
> CoreTiming::GetGlobalTimeUsScaled()
> ```
> 
> But maybe the rest of the differences is fixing the adhocMatching you mentioned that's broken in PPSSPP's current master. Feel free to salvage that code, it's yours anyway.
> 
> You could also take my changes for README.md where I removed duplicate credits we already had in it's proper place and UI/MiscScreens.cpp where I added you to the credits and aemu to the 3rd party tools used to keep it consistent with everything else and finally place your name there at the same time.

Although CoreTiming::GetGlobalTimeUsScaled() vs real_time_now()*1000000.0 may affect stability (ie. disconnection issue) but it shouldn't cause the game to freeze for an indefinitely time, that's why i suspects there is a deadlock/race-condition, but i haven't debug the code yet, because i need to make sure multiple instance feature to work properly first to make testing adhoc easier on a single laptop.

Btw, how do i exclude Android from multiple instance feature? since android doesn't have **shm_open** and may need to replace it with **ashmem** which is a hassle since i'm not familiar with it (i don't even know what to imports)

anr2me committed:
Oops forgot to change this also for multiple-instance to work properly

anr2me committed:
Excluding Android from multiple-instance feature because i'm too lazy to learn ashmem :D

ANR2ME commented:
Btw, currently on every Reset/changing game the PPSSPP_ID keeps increasing thus having different IP, so the adhoc server need to be re-pointed to the right IP on every reset/changing game which can be a hassle. 

It would be nice if the PPSSPP_ID can be made persistent but i forgot how :) i think the PPSSPP_ID Initialization & Cleanup need to be moved somewhere else

ANR2ME commented on Core/HLE/proAdhoc.cpp:
```diff
@@ -1296,11 +1304,26 @@ int getLocalIp(sockaddr_in * SocketAddress){
+		}
+		return 0;
+	}
+	//return -1;
```
I prefer to return -1 (failed) if the address didn't passed the validity test tho, but since i'm not trying to be in the way of the person who wrote the previous code so i just commented my code :D just in case that person have plans for his code in the future

LunaMoo commented:
Please avoid looking at https://github.com/LunaMoo/ppsspp/commit/4ec7f4f26b3ece66b64d8b2830a4b16718f55682 and instead look at the branch I compared to which is https://github.com/LunaMoo/ppsspp/commits/AdhocStuff_from_THBranch, all the adhoc stuff before adding hacks is in this commit https://github.com/LunaMoo/ppsspp/commit/c9de850c7d597dc32c5c064b3afedf3a0942672b 

They're different as I only started testing adhoc on my TemporaryHackishBranch later on and fixed some stuff I missed on the way that were missing in the initial merge while this separate branch was made specially in case someone wanted to merge multiply instances functionality as well as somewhat improved compatibility and just included all currently used adhoc code.

ANR2ME commented:
> 
> 
> Please avoid looking at [LunaMoo@4ec7f4f](https://github.com/LunaMoo/ppsspp/commit/4ec7f4f26b3ece66b64d8b2830a4b16718f55682) and instead look at the branch I compared to which is https://github.com/LunaMoo/ppsspp/commits/AdhocStuff_from_THBranch, all the adhoc stuff before adding hacks is in this commit [LunaMoo@c9de850](https://github.com/LunaMoo/ppsspp/commit/c9de850c7d597dc32c5c064b3afedf3a0942672b)
> 
> They're different as I only started testing adhoc on my TemporaryHackishBranch later on and fixed some stuff I missed on the way that were missing in the initial merge while this separate branch was made specially in case someone wanted to merge multiply instances functionality as well as somewhat improved compatibility and just included all currently used adhoc code.

Currently i'm trying to separate pull request of multiple-instance feature from adhoc bug fix.
I'll create a different pull request for things related to adhoc compatibility/stability when i found a solution, as i remembered the last time i worked with AdhocMatching the callbacks is corrupting some registers when returned and some even failed to return from callbacks (within 250ms) because the callback runs on any random PSP Thread (current thread). 
I did some experiment last night, testing to run AdhocMatching callbacks on idleThread(by calling __KernelSwitchOffThread before __KernelDirectMipsCall) seems to be the one failed to return within 250ms.
And currently i'm experimenting to run the mipscall on AdhocThread but having difficulty converting SceUID to Thread to be used on __KernelSwitchContext before calling __KernelDirectMipsCall

This is the code i'm having issue and don't know how to solve it: Error C2440	'static_cast': cannot convert from 'KernelObject *' to 'T *'
`Thread *t = kernelObjects.GetFast<Thread>(threadAdhocID);`

PS: On JPCSP when AdhocMatching Started it will create a new Thread, even though this thread is just a dummy/loop that calls HLE function (the actual code is running on JPCSP side) at least mipscall will runs on it's own thread thus won't corrupt registers on any random thread.

ANR2ME commented:
i probably should rebase this in to a single commit, and the branch name was suppose for fixing bugs related to adhoc instead of feature addition >.< i should change the branch name also
